rsocket/RSocketClient.cpp

```diff
@@ -118,15 +124,61 @@ folly::Future<folly::Unit> RSocketClient::resume() {
   });
 }
 
+folly::Future<folly::Unit> RSocketClient::close(
+    folly::exception_wrapper ew,
+    StreamCompletionSignal signal) {
+  if (disconnectBaton_) {
+    disconnectBaton_->wait();
+    disconnectBaton_.clear();
+  }
+  disconnectBaton_.emplace();
+
+
+  // Immediately move the stateMachine_, so next calls will not fail on it.
+  auto sm = std::move(stateMachine_);
+
+  if (!sm) {
+    return folly::makeFuture<folly::Unit>(
+        std::runtime_error{"RSocketClient must always have a state machine"});
+  }
+
+  auto work = [ sm = sm, e = std::move(ew), signal, this ]() mutable {
+    sm->close(std::move(e), signal);
+    disconnectBaton_->post();
+  };
+
+  if (evb_->isInEventBaseThread()) {
+    VLOG(2) << "Running RSocketClient close synchronously";
+    work();
+    return folly::unit;
+  }
+
+  VLOG(2) << "Scheduling RSocketClient close";
+  return folly::via(evb_, work);
+}
+
 folly::Future<folly::Unit> RSocketClient::disconnect(
     folly::exception_wrapper ew) {
   if (!stateMachine_) {
     return folly::makeFuture<folly::Unit>(
         std::runtime_error{"RSocketClient must always have a state machine"});
   }
 
-  auto work = [ sm = stateMachine_, e = std::move(ew) ]() mutable {
+  if (!stateMachine_->isResumable()) {
+    LOG(INFO) << "Not resumable, so close the socket!";
+    return close(ew, StreamCompletionSignal::SOCKET_CLOSED);
+  }
+
+  if (disconnectBaton_) {
+    disconnectBaton_->wait();
+    disconnectBaton_.clear();
+  }
+
+  disconnectBaton_.emplace();
+
+  auto work = [ sm = stateMachine_, e = std::move(ew), this ]() mutable {
```

Comment:
you can't pass this inside of the lambda, Dylan just fixed an issue where we were accessing "this" which was already gone.